src/navier.f90

```diff
@@ -1280,5 +1289,246 @@ subroutine tbl_flrt (ux1,uy1,uz1)
     enddo
 
   end subroutine tbl_flrt
+!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!
+  !!
+  !!  subroutine: pipe_bulk / pipe_bulk_u / pipe_bulk_phi
+  !!      AUTHOR: Rodrigo Vicente Cruz
+  !! DESCRIPTION: Correction of pipe's bulk velocity (constant 
+  !!              flow rate) and bulk temperature.
+  !!              See Thesis Vicente Cruz 2021 for help.
+  !!
+!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!
+  !********************************************************************
+  !
+  subroutine pipe_bulk(ux,uy,uz,ep)
+  !
+  !********************************************************************
+
+    use param
+    use decomp_2d
+    use variables
+    use var,  only: phi1
+
+    implicit none
+    real(mytype),dimension(xsize(1),xsize(2),xsize(3))  :: ux,uy,uz,ep
+    integer                                             :: is
+
+    !Bulk velocity correction
+    call pipe_bulk_u(ux,uy,uz,ep,one)
+
+    !Bulk temperature correction
+    if (numscalar.ne.0) then
+        do is=1,numscalar
+            call pipe_bulk_phi(phi1(:,:,:,is),ux,ep,is,one)
+        enddo
+    endif
+
+  end subroutine pipe_bulk
+  !********************************************************************
+  !
+  subroutine pipe_bulk_u(ux,uy,uz,ep,ub_constant)
+  !
+  !********************************************************************
+
+    use decomp_2d
+    use variables
+    use param
+    use var
+    use ibm_param, only: rai
+    use MPI
+
+    implicit none
+    !INPUTS
+    real(mytype),intent(inout),dimension(xsize(1),xsize(2),xsize(3))    :: ux,uy,uz
+    real(mytype),intent(in   ),dimension(xsize(1),xsize(2),xsize(3))    :: ep
+    real(mytype),intent(in   )                                          :: ub_constant !bulk velocity value
+    !LOCALS
+    real(mytype)                                                        :: qm      !flow rate
+    real(mytype)                                                        :: ym,zm,yc,zc,r
+    real(mytype)                                                        :: ncount
+    integer                                                             :: j,i,k,code
+    character(len=30)                                                   :: filename
+    integer, save                                                       :: local_io_unit=-1
+
+    yc = yly/two
+    zc = zlz/two
+
+    if (local_io_unit.eq.-1 .and. nrank.eq.0) then
+       open(newunit=local_io_unit,file='Ub.dat',status='unknown')
+    endif
+
+    !--------------------------- Bulk Velocity ---------------------------
+    !Calculate loss of streamwise mean pressure gradient
+    call pipe_volume_avg(ux,qm,ep)
+    if (nrank==0) then
+       if (mod(itime, ilist)==0) print *,'Velocity:'
+       if (mod(itime, ilist)==0) print *,'    Bulk velocity before',qm
+       write(local_io_unit,*) real((itime-1)*dt,mytype), (ub_constant-qm) !write pressure drop
+    endif
+
+    !Correction
+    do j=1,xsize(2)
```

Comment:
I think the loop order for efficient memory access should be
```
do k = 
   do j = 
      do i =
```
But the current order is faster in the IBM framework ?

Reply:
Thank you for spotting that, it was indeed a mistake passed on from my old Incompact3d version, where _y_ was the streamwise direction and not _x_. Anyway, this has been corrected in 25d46cc43400701dbc33aadb5d8c25638838e334.